xml/art_virtualization-best-practices.xml

```diff
@@ -3241,6 +3241,162 @@ LC_ALL=C PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin \
+            <literal>path</literal> with value to the socket, and finally
+            <literal>mode</literal> which accepts the unique valid value
+            <literal>client</literal> specifying the role of hypervisor. It's
+            recommended to allow libvirt manage the persistent reservations.
+          </para>
+        </callout>
+        <callout arearefs="co-disk-scsi-hba-unit">
+          <para>
+            please make sure the virtio-scsi HBA that the disk attaches already
+            exists and has available units(maximum count of units per virtio-scsi
+            HBA is 7), Otherwise you need to manually add a virtio-scsi HBA to
+            avoid automatically adding the LSI HBA by libvirt. For example:
+<screen>&lt;controller type='scsi' index='0' model='virtio-scsi'&gt;
+  &lt;address type='pci' domain='0x0000' bus='0x03' slot='0x00' function='0x0'/&gt;
+&lt;/controller&gt;</screen>
+          </para>
+        </callout>
+      </calloutlist>
+<screen>&prompt.user;&sudo; virsh domblklist sles15sp5
+ Target   Source
+---------------------------------------------
+ vda      /mnt/images/sles15sp5/disk0.qcow2
+ sda      /dev/mapper/storage1</screen>
+      <para>
+        Start the &vmguest; then perform following commands In &vmguest;:
+      </para>
+<screen>&prompt.user;lsblk
+NAME   MAJ:MIN RM  SIZE RO TYPE MOUNTPOINTS
+sda      8:0    0   50G  0 disk
+vda    253:0    0   20G  0 disk
+├─vda1 253:1    0    8M  0 part
+├─vda2 253:2    0    2G  0 part [SWAP]
+└─vda3 253:3    0   18G  0 part /</screen>
+      <para>
+        Let's reserve the scsi disk with the key <literal>123abc</literal>, then
+        release it:
+      </para>
+<screen>&prompt.user;&sudo; sg_persist --verbose --out --register --param-sark=123abc /dev/sda
+    inquiry cdb: [12 00 00 00 24 00]
+  TrueNAS   iSCSI Disk        0123
+  Peripheral device type: disk
+    Persistent reservation out cdb: [5f 00 00 00 00 00 00 00 18 00]
+PR out: command (Register) successful</screen>
+<screen>&prompt.user;&sudo; sg_persist --verbose --in -k /dev/sda
+    inquiry cdb: [12 00 00 00 24 00]
+  TrueNAS   iSCSI Disk        0123
+  Peripheral device type: disk
+    Persistent reservation in cdb: [5e 00 00 00 00 00 00 20 00 00]
+  PR generation=0x5, 2 registered reservation keys follow:
+    0x123abc
+    0x123abc</screen>
+<screen>&prompt.user;&sudo; sg_persist --verbose --out --clear --param-rk=123abc /dev/sda
+    inquiry cdb: [12 00 00 00 24 00]
+  TrueNAS   iSCSI Disk        0123
+  Peripheral device type: disk
+    Persistent reservation out cdb: [5f 03 00 00 00 00 00 00 18 00]
+PR out: command (Clear) successful</screen>
+<screen>&prompt.user;&sudo; sg_persist --verbose --in -k /dev/sda
+    inquiry cdb: [12 00 00 00 24 00]
+  TrueNAS   iSCSI Disk        0123
+  Peripheral device type: disk
+    Persistent reservation in cdb: [5e 00 00 00 00 00 00 20 00 00]
+  PR generation=0x6, there are NO registered reservation keys</screen>
+    </sect2>
```

Comment:
Below are the initial qemu commits about qemu-pr-helper:

commit fe8fc5ae5c808e037fa4746cbfeb3c07ffe0af81
Author: Paolo Bonzini <pbonzini@redhat.com>
Date:   Tue Aug 22 06:50:55 2017 +0200

    scsi: add multipath support to qemu-pr-helper
    
    Proper support of persistent reservation for multipath devices requires
    communication with the multipath daemon, so that the reservation is
    registered and applied when a path comes up.  The device mapper
    utilities provide a library to do so; this patch makes qemu-pr-helper.c
    detect multipath devices and, when one is found, delegate the operation
    to libmpathpersist.
    
    Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

commit b855f8d175a0a26c9798cbc5962bb8c0d9538231
Author: Paolo Bonzini <pbonzini@redhat.com>
Date:   Tue Aug 22 06:50:18 2017 +0200

    scsi: build qemu-pr-helper
    
    Introduce a privileged helper to run persistent reservation commands.
    This lets virtual machines send persistent reservations without using
    CAP_SYS_RAWIO or out-of-tree patches.  The helper uses Unix permissions
    and SCM_RIGHTS to restrict access to processes that can access its socket
    and prove that they have an open file descriptor for a raw SCSI device.
    
    The next patch will also correct the usage of persistent reservations
    with multipath devices.
    
    It would also be possible to support for Linux's IOC_PR_* ioctls in
    the future, to support NVMe devices.  For now, however, only SCSI is
    supported.
    
    Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
    
It seems that above commit logs can satisfy your concern more or less.
But if I copied/pasted them to the document, I'm afraid that you'll ask for adding all of detailed explanation about all of information that introduced by commit logs.
I'm not the author of qemu-pr-helper, I'm not a storage expert, not familiar with the details of scsi persistent reservation either.

It seems that we need a storage expert getting involve to help to finish this document.

Reply:
I think this can be skipped completely, we are trying to tell a sysadmin what to do in practice to achieve the goal,
we are not explaining to the sysadmin how SCSI protocol works or which qemu processes and qemu helpers are spawned in the background to accomplish the tasks, this is irrelevant from a user perspective I think.